sdk/azcore/arm/resource_identifier.go

```diff
@@ -0,0 +1,321 @@
+//go:build go1.16
+// +build go1.16
+
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package arm
+
+import (
+	"fmt"
+	"strings"
+)
+
+const (
+	providersKey             = "providers"
+	subscriptionsKey         = "subscriptions"
+	resourceGroupsLowerKey   = "resourcegroups"
+	locationsKey             = "locations"
+	builtInResourceNamespace = "Microsoft.Resources"
+)
+
+var (
+	// RootResourceIdentifier defines a ResourceIdentifier of a tenant as a root level parent of all other ResourceIdentifier
+	RootResourceIdentifier = &ResourceIdentifier{
+		parent:       nil,
+		resourceType: TenantResourceType,
+		name:         "",
+	}
+
+	// SubscriptionResourceType is the ResourceType of a subscription
+	SubscriptionResourceType  = NewResourceType("Microsoft.Resources", "subscriptions")
+	// ResourceGroupResourceType is the ResourceType of a resource group
+	ResourceGroupResourceType = NewResourceType("Microsoft.Resources", "resourceGroups")
+	// TenantResourceType is the ResourceType of a tenant
+	TenantResourceType        = NewResourceType("Microsoft.Resources", "tenants")
+	// ProviderResourceType is the ResourceType of a provider
+	ProviderResourceType = NewResourceType("Microsoft.Resources", "providers")
+)
+
+// ResourceIdentifier represents a resource ID such as `/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/myRg`
+type ResourceIdentifier struct {
+	parent            *ResourceIdentifier
+	subscriptionId    string
+	provider          string
+	resourceGroupName string
+	location          string
+	resourceType      ResourceType
+	name              string
+	isChild           bool
+
+	stringValue string
+}
+
+func (id ResourceIdentifier) Parent() *ResourceIdentifier {
+	return id.parent
+}
+
+func (id ResourceIdentifier) SubscriptionId() string {
```

Comment:
Since the `Parse()` method returns `ResourceIdentifier` by reference, you should make the receivers also by reference.

Reply:
I have made the receivers of ResourceIdentifier and ResourceType pointers